cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -466,10 +477,21 @@ impl MachInstEmit for Inst {
                 debug_assert_eq!(rs1, a0());
                 debug_assert_eq!(rs2, b0());
                 let rd = allocs.next_writable(rd);
-                put_string(
-                    &format!("$ => {} :{}\n", reg_name(rd.to_reg()), alu_op.op_name()),
-                    sink,
-                );
+                // it is discussable if MulArith should be separate inst or part of ALURRR
+                // also, should I somehow show that we are using C register here?
```

Comment:
Yeah, AluRRR is not at all appropriate in this instance. This first, and foremost, is something RISC-V specific. It has ALU instructions with three register arguments, which all can be arbitrary GPRs: two inputs and one output. Hence the RRR.

`:ARITH` instruction, on the other hand does not provide that flexibility in register allocation. It requires that the arguments are placed in the very specific registers before it executes, and these registers are *implicit* uses, and there are 3 of them, so what you’d really need is actually `AluRRRR` if we were going for abstracting away this pattern. This mismatch is also why you're finding yourself clobbering the C register in the implementation here – doing so is not valid as the register allocator might have assigned a value to C earlier and could be expecting this value to survive an execution of the `ARITH` instruction.

I would say that in general if you encounter artifacts from the RISC-V world, treat them with suspicion and default to removing and replacing them with something fresh.